qiskit/circuit/controlflow/if_else.py

```diff
@@ -484,10 +557,27 @@ def _unify_circuit_bits_rebuild(
     qubits = list(set(true_body.qubits).union(false_body.qubits))
     clbits = list(set(true_body.clbits).union(false_body.clbits))
     # We use the inner `_append` method because everything is already resolved.
-    true_out = QuantumCircuit(qubits, clbits)
+    true_out = QuantumCircuit(qubits, clbits, *true_body.qregs, *true_body.cregs)
     for data in true_body.data:
         true_out._append(*data)
-    false_out = QuantumCircuit(qubits, clbits)
+    false_out = QuantumCircuit(qubits, clbits, *false_body.qregs, *false_body.cregs)
     for data in false_body.data:
         false_out._append(*data)
-    return true_out, false_out
+    return _unify_circuit_registers(true_out, false_out)
+
+
+def _unify_circuit_registers(
+    true_body: QuantumCircuit, false_body: QuantumCircuit
+) -> Tuple[QuantumCircuit, QuantumCircuit]:
+    """
+    Ensure that ``true_body`` and ``false_body`` have the same registers defined within them.  These
+    do not need to be in the same order between circuits.  The two input circuits are returned,
+    mutated to have the same registers.
+    """
+    true_registers = set(true_body.qregs) | set(true_body.cregs)
+    false_registers = set(false_body.qregs) | set(false_body.cregs)
+    for register in false_registers - true_registers:
+        true_body.add_register(register)
```

Comment:
Ah, I'm disappointed that `add_register` doesn't gracefully handle duplicate additions.

Reply:
It's also linear in the number of registers already in the circuit :(